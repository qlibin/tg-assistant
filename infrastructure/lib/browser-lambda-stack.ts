import { Duration, Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ecr from 'aws-cdk-lib/aws-ecr';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';

export interface BrowserLambdaStackProps extends StackProps {
  environmentName: string;
  ecrRepoName: string;
  lambdaName: string;
  imageTag: string; // default provided by app
  tags?: Record<string, string>;
}

export class BrowserLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: BrowserLambdaStackProps) {
    super(scope, id, props);

    const { environmentName, ecrRepoName, lambdaName, imageTag } = props;

    // Lookup existing ECR repository by name (account/region from env)
    const repository = ecr.Repository.fromRepositoryName(this, 'Repository', ecrRepoName);

    // Execution role for Lambda
    const execRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `browser-lambda-role-${environmentName}`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Execution role for browser-lambda to pull from ECR and write logs',
    });

    execRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')
    );

    // Allow pull from ECR with least privilege
    execRole.addToPolicy(
      new iam.PolicyStatement({
        actions: ['ecr:GetAuthorizationToken'],
        resources: ['*'],
      })
    );
    execRole.addToPolicy(
      new iam.PolicyStatement({
        actions: [
          'ecr:BatchCheckLayerAvailability',
          'ecr:GetDownloadUrlForLayer',
          'ecr:BatchGetImage',
        ],
        resources: [repository.repositoryArn],
      })
    );

    // Image URI with tag
    const imageUri = `${Stack.of(this).account}.dkr.ecr.${Stack.of(this).region}.amazonaws.com/${ecrRepoName}:${imageTag}`;

    const fn = new lambda.DockerImageFunction(this, 'Function', {
      functionName: lambdaName,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 1024,
      timeout: Duration.seconds(300),
      role: execRole,
      code: lambda.DockerImageCode.fromEcr(repository, {
        tagOrDigest: imageTag,
      }),
      environment: {
        NODE_ENV: 'production',
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    new CfnOutput(this, 'FunctionName', { value: fn.functionName });
    new CfnOutput(this, 'FunctionArn', { value: fn.functionArn });
    new CfnOutput(this, 'ImageUri', { value: imageUri });
  }
}
