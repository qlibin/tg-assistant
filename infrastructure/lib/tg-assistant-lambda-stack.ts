import { Annotations, CfnOutput, Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as secretsmanager from 'aws-cdk-lib/aws-secretsmanager';

export interface TgAssistantLambdaStackProps extends StackProps {
  environmentName: string;
  lambdaName: string;
  tags?: Record<string, string>;
}

export class TgAssistantLambdaStack extends Stack {
  constructor(scope: Construct, id: string, props: TgAssistantLambdaStackProps) {
    super(scope, id, props);

    const { environmentName, lambdaName } = props;

    // Execution role for Lambda (least privilege: basic execution role)
    const execRole = new iam.Role(this, 'LambdaExecutionRole', {
      roleName: `telegram-webhook-lambda-role-${environmentName}`,
      assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
      description: 'Execution role for Lambda to write logs',
    });

    execRole.addManagedPolicy(
      iam.ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSLambdaBasicExecutionRole')
    );

    // Choose code source: use pre-built ZIP when provided by CI, otherwise inline minimal handler to allow synth/tests without filesystem assets
    const lambdaZipPath = process.env.LAMBDA_ZIP_PATH;
    const useInline = !lambdaZipPath;

    const code = useInline
      ? lambda.Code.fromInline(
          'exports.handler = async () => { return { statusCode: 200, body: "ok" }; };'
        )
      : lambda.Code.fromAsset(lambdaZipPath);

    const handler = useInline ? 'index.handler' : 'dist/index.handler';

    // Create or reference the Telegram webhook secret per environment
    const secretName = `/tg-assistant/telegram-webhook-secret/${environmentName}`;
    const telegramWebhookSecret = new secretsmanager.Secret(this, 'TelegramWebhookSecret', {
      secretName,
      description: 'Telegram webhook secret used to validate updates',
      generateSecretString: {
        // Placeholder; value should be updated out-of-band as per runbook
        secretStringTemplate: JSON.stringify({}),
        generateStringKey: 'webhookSecret',
        passwordLength: 32,
        excludePunctuation: true,
      },
    });

    const fn = new lambda.Function(this, 'Function', {
      functionName: lambdaName,
      runtime: lambda.Runtime.NODEJS_22_X,
      architecture: lambda.Architecture.ARM_64,
      memorySize: 1024,
      timeout: Duration.seconds(300),
      role: execRole,
      code,
      handler,
      environment: {
        NODE_ENV: 'production',
        TELEGRAM_WEBHOOK_SECRET_ARN: telegramWebhookSecret.secretArn,
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
    });

    // Grant Lambda permission to read the secret value
    telegramWebhookSecret.grantRead(fn);

    // Context parameter for external API Gateway invoke permission (full SourceArn provided)
    const sourceArnRaw = this.node.tryGetContext('apiGatewaySourceArn') as unknown;
    const sourceArn =
      typeof sourceArnRaw === 'string' && sourceArnRaw.trim().length > 0
        ? sourceArnRaw.trim()
        : undefined;

    if (sourceArn) {
      new lambda.CfnPermission(this, 'ApiGatewayInvokePermission', {
        action: 'lambda:InvokeFunction',
        functionName: fn.functionArn,
        principal: 'apigateway.amazonaws.com',
        sourceArn,
      });
    } else {
      Annotations.of(this).addWarning(
        'API Gateway SourceArn not provided. Skipping Lambda invoke permission. Provide context: -c apiGatewaySourceArn=arn:aws:execute-api:...'
      );
    }

    new CfnOutput(this, 'FunctionName', { value: fn.functionName });
    new CfnOutput(this, 'FunctionArn', { value: fn.functionArn });
    new CfnOutput(this, 'LambdaRegion', { value: Stack.of(this).region });
    if (sourceArn) {
      new CfnOutput(this, 'ApiGatewaySourceArn', { value: sourceArn });
    }
    new CfnOutput(this, 'TelegramWebhookSecretArn', { value: telegramWebhookSecret.secretArn });
  }
}
