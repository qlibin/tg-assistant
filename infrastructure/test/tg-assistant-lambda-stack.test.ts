import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as cdk from 'aws-cdk-lib';
import { Match, Template } from 'aws-cdk-lib/assertions';
import { TgAssistantLambdaStack } from '../lib/tg-assistant-lambda-stack';

// AAA pattern tests for CDK stack

describe('TgAssistantLambdaStack (ZIP-based Node.js Lambda)', () => {
  const baseEnv = { account: '123456789012', region: 'us-east-1' } as const;

  const makeStack = (
    overrides?: Partial<{
      envName: string;
      lambdaName: string;
      setZipPath: boolean;
    }>
  ) => {
    const app = new cdk.App();

    if (overrides?.setZipPath) {
      // Provide a directory path as asset source so CDK can hash it without requiring a real ZIP
      process.env.LAMBDA_ZIP_PATH = fs.mkdtempSync(path.join(os.tmpdir(), 'asset-'));
    } else {
      delete process.env.LAMBDA_ZIP_PATH;
    }

    return new TgAssistantLambdaStack(app, 'TestStack', {
      env: baseEnv,
      description: 'Test stack',
      environmentName: overrides?.envName ?? 'dev',
      lambdaName: overrides?.lambdaName ?? 'telegram-webhook-lambda-dev',
      tags: { app: 'telegram-webhook', env: overrides?.envName ?? 'dev' },
    });
  };

  test('synthesizes expected CloudFormation template (snapshot)', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev', setZipPath: true });

    // Act
    const templateJson = Template.fromStack(stack).toJSON();

    // Assert
    expect(templateJson).toMatchSnapshot();
  });

  test('creates Secrets Manager secret per environment and exposes ARN to Lambda env', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev', setZipPath: true });

    // Act
    const template = Template.fromStack(stack);

    // Assert Secret resource
    template.hasResourceProperties('AWS::SecretsManager::Secret', {
      Name: '/tg-assistant/telegram-webhook-secret/dev',
      Description: 'Telegram webhook secret used to validate updates',
    });

    // Assert Lambda has env var wired to secret ARN (token acceptable)
    template.hasResourceProperties(
      'AWS::Lambda::Function',
      Match.objectLike({
        Environment: Match.objectLike({
          Variables: Match.objectLike({
            TELEGRAM_WEBHOOK_SECRET_ARN: Match.anyValue(),
          }),
        }),
      })
    );

    // Assert output for Secret ARN exists
    template.hasOutput('TelegramWebhookSecretArn', Match.anyValue());
  });

  test('creates Lambda with expected config (runtime, arch, memory, timeout, env)', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev', setZipPath: true });

    // Act
    const template = Template.fromStack(stack);

    // Assert
    template.hasResourceProperties(
      'AWS::Lambda::Function',
      Match.objectLike({
        FunctionName: 'telegram-webhook-lambda-dev',
        Runtime: 'nodejs22.x',
        MemorySize: 1024,
        Timeout: 300,
        Architectures: ['arm64'],
        Environment: Match.objectLike({
          Variables: Match.objectLike({
            NODE_ENV: 'production',
            TELEGRAM_WEBHOOK_SECRET_ARN: Match.anyValue(),
          }),
        }),
        Handler: 'dist/index.handler',
      })
    );
  });

  test('execution role trusts Lambda service and has AWSLambdaBasicExecutionRole', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev' });
    const template = Template.fromStack(stack);

    // Assert trust policy
    template.hasResourceProperties('AWS::IAM::Role', {
      AssumeRolePolicyDocument: {
        Statement: Match.arrayWith([
          Match.objectLike({
            Principal: { Service: 'lambda.amazonaws.com' },
            Action: 'sts:AssumeRole',
          }),
        ]),
      },
    });

    // Assert managed policy attachment via ManagedPolicyArns on the Role
    template.hasResourceProperties('AWS::IAM::Role', {
      ManagedPolicyArns: Match.anyValue(),
    });
  });

  test('configures log retention for 30 days', () => {
    // Arrange
    const stack = makeStack({ envName: 'dev' });
    const template = Template.fromStack(stack);

    // Assert
    template.hasResourceProperties('Custom::LogRetention', {
      RetentionInDays: 30,
    });
  });
});
