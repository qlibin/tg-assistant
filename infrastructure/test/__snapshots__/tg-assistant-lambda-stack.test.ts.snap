// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TgAssistantLambdaStack (ZIP-based Node.js Lambda) creates Lambda invoke permission for HTTP API (pattern covered by snapshot) 1`] = `
{
  "Description": "Test stack",
  "Outputs": {
    "ApiGatewaySourceArn": {
      "Value": "arn:aws:execute-api:us-east-1:123456789012:abc123/beta/*",
    },
    "FunctionArn": {
      "Value": {
        "Fn::GetAtt": [
          "Function76856677",
          "Arn",
        ],
      },
    },
    "FunctionName": {
      "Value": {
        "Ref": "Function76856677",
      },
    },
    "LambdaRegion": {
      "Value": "us-east-1",
    },
    "TelegramWebhookSecretArn": {
      "Value": {
        "Ref": "TelegramWebhookSecret2F3DC73E",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ApiGatewayInvokePermission": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "Function76856677",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": "arn:aws:execute-api:us-east-1:123456789012:abc123/beta/*",
      },
      "Type": "AWS::Lambda::Permission",
    },
    "Function76856677": {
      "DependsOn": [
        "LambdaExecutionRoleDefaultPolicy6D69732F",
        "LambdaExecutionRoleD5C26073",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "ZipFile": "exports.handler = async () => { return { statusCode: 200, body: "ok" }; };",
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
            "TELEGRAM_WEBHOOK_SECRET_ARN": {
              "Ref": "TelegramWebhookSecret2F3DC73E",
            },
          },
        },
        "FunctionName": "telegram-webhook-lambda-dev",
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRoleD5C26073",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionLogRetention5FDF6B4D": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "Function76856677",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaExecutionRoleD5C26073": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Execution role for Lambda to write logs",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "telegram-webhook-lambda-role-dev",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaExecutionRoleDefaultPolicy6D69732F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "TelegramWebhookSecret2F3DC73E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaExecutionRoleDefaultPolicy6D69732F",
        "Roles": [
          {
            "Ref": "LambdaExecutionRoleD5C26073",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TelegramWebhookSecret2F3DC73E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "Telegram webhook secret used to validate updates",
        "GenerateSecretString": {
          "ExcludePunctuation": true,
          "GenerateStringKey": "webhookSecret",
          "PasswordLength": 32,
          "SecretStringTemplate": "{}",
        },
        "Name": "/tg-assistant/telegram-webhook-secret/dev",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`TgAssistantLambdaStack (ZIP-based Node.js Lambda) synthesizes expected CloudFormation template (snapshot) 1`] = `
{
  "Description": "Test stack",
  "Outputs": {
    "FunctionArn": {
      "Value": {
        "Fn::GetAtt": [
          "Function76856677",
          "Arn",
        ],
      },
    },
    "FunctionName": {
      "Value": {
        "Ref": "Function76856677",
      },
    },
    "LambdaRegion": {
      "Value": "us-east-1",
    },
    "TelegramWebhookSecretArn": {
      "Value": {
        "Ref": "TelegramWebhookSecret2F3DC73E",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "Function76856677": {
      "DependsOn": [
        "LambdaExecutionRoleDefaultPolicy6D69732F",
        "LambdaExecutionRoleD5C26073",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "1ebc9d3ac2033816c4abb63e4afd69d350b4aba8704cc9236b82ea520b74f4b0.zip",
        },
        "Environment": {
          "Variables": {
            "NODE_ENV": "production",
            "TELEGRAM_WEBHOOK_SECRET_ARN": {
              "Ref": "TelegramWebhookSecret2F3DC73E",
            },
          },
        },
        "FunctionName": "telegram-webhook-lambda-dev",
        "Handler": "dist/index.handler",
        "MemorySize": 1024,
        "Role": {
          "Fn::GetAtt": [
            "LambdaExecutionRoleD5C26073",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "FunctionLogRetention5FDF6B4D": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "Function76856677",
              },
            ],
          ],
        },
        "RetentionInDays": 30,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaExecutionRoleD5C26073": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": "Execution role for Lambda to write logs",
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "RoleName": "telegram-webhook-lambda-role-dev",
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaExecutionRoleDefaultPolicy6D69732F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "secretsmanager:GetSecretValue",
                "secretsmanager:DescribeSecret",
              ],
              "Effect": "Allow",
              "Resource": {
                "Ref": "TelegramWebhookSecret2F3DC73E",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaExecutionRoleDefaultPolicy6D69732F",
        "Roles": [
          {
            "Ref": "LambdaExecutionRoleD5C26073",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-123456789012-us-east-1",
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "TelegramWebhookSecret2F3DC73E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "Telegram webhook secret used to validate updates",
        "GenerateSecretString": {
          "ExcludePunctuation": true,
          "GenerateStringKey": "webhookSecret",
          "PasswordLength": 32,
          "SecretStringTemplate": "{}",
        },
        "Name": "/tg-assistant/telegram-webhook-secret/dev",
      },
      "Type": "AWS::SecretsManager::Secret",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
